aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -346,27 +330,76 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
         }
     );
 
+    if (r.scalar_type() != commonDtype) {
+      r_values = r_values.to(r.scalar_type());
+    }
+    get_sparse_impl(r)->set_indices_and_values_unsafe(r_indices, r_values);
     get_sparse_impl(r)->set_nnz_and_narrow(r_i);
+
     // TODO: I think it may be possible to track inside the loop and
     // detect when we are uncoalesced (e.g., by observing that an
     // index goes backwards) which may be more precise than using the
     // coalesced flag here.  But this is easy.
-    return r._coalesced_(t_coalesced && s_coalesced);
-  } else {
+    return r._coalesced_(coalesced);
+}
+
+SparseTensor& add_out_sparse_non_contiguous(SparseTensor& r, const SparseTensor& t, const SparseTensor& src, Scalar value, ScalarType commonDtype) {
+    Tensor t_values = maybe_promoted_tensor(t._values(), commonDtype);
+    Tensor s_values = maybe_promoted_tensor(src._values(), commonDtype);
+
     // If `t` or `src` contains non-contiguous `values`, `THBlas_axpy` doesn't work
     // and we concat the indices and values tensors instead.
     AT_DISPATCH_ALL_TYPES(
-      s_values.scalar_type(), "add_out_sparse_cuda", [&] {
+      commonDtype, "add_out_sparse_cuda", [&] {
           if (value.to<scalar_t>() != static_cast<scalar_t>(1)) {
             s_values = s_values.mul(value);
           }
         });
 
-    LongTensor r_indices = at::cat({t_indices, src_indices}, 1);
+    LongTensor r_indices = at::cat({t._indices(), src._indices()}, 1);
     Tensor r_values = at::cat({t_values, s_values}, 0);
+
+    if (r.scalar_type() != commonDtype) {
+      r_values = r_values.to(r.scalar_type());
+    }
     alias_into_sparse(r, r_indices, r_values);
 
     return r;
+}
+
+Tensor& add_out_dense_sparse_cpu(Tensor& r, const Tensor& dense, const SparseTensor& sparse_, Scalar value);
+
+SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const SparseTensor& src, Scalar value) {
+  if (!t.is_sparse()) {
+    return add_out_dense_sparse_cpu(r, t, src, value);
```

Comment:
I haven't read through all of the code, but right now, `add_out_dense_sparse_cpu` handles type promotion internally. How difficult would it be to just handle type promotion just once in `add_out_sparse_cpu` for all of the add functions?

Reply:
I think similarly to my comment on the previous diff, I can see why breaking out just the type promotion into a wrapper function would make the diff easier to read, but I'm not sure it makes the end-result of the code simpler.

add_out_sparse_cpu does all of the error checking and setup to perform the add operation (pulling out the indices/values, handling special cases like nnz==0, etc, and then calls a function to do the addition. Of all the setup steps needed, type promotion doesn't seem to me to be special in a way that separating it from the rest of the setup would make the code clearer. Many levels of nesting could potentially make it less clear.

I think if I was to refactor further I'd just want to break apart ops like `mul` to be more like this add function -- I'd like a wrapper that does all of the setup and an impl that does everything else, not necessarily a wrapper to do promotion wrapping a wrapper that does the error checking and rest of the setup that wraps an impl.

Reply:
I think you might have misinterpreted my comment's intent. It is not so much about making the diff easier to review (although that would be nice), but I'm worried about the readability of the code in the long term. Here's an anecdote:

Someone (potentially new to the project) comes along and asks, "how does type promotion in pytorch work?" (That is the question I am asking as I read this diff). They look at the TensorIterator code and see that it is implemented in exactly the same way for add, sub, mul, div. Good. Next, they look at how sparse arithmetic ops are implemented. Instead of having one entry point for type promotion, a function like `add` has ~4 different places that do type promotion separately. This is confusing.

Reply:
Without considering promotion, sparse math generally starts at native_functions.yaml pointing to functions that:

1. Do error checking and may short circuit to another impl depending on layout of params
2. Extract indices/values/nnz/etc on which to operate, possibly construct an empty result.
3. Do the math or call a function to do the math
4. Put the indices/values back into the (possibly in-place) result and return it.

>  a function like add has ~4 different places that do type promotion separately.

I think this just means the four calls (add_sparse, add_sparse_, add_out_sparse_cpu, and add_out_sparse_cuda) dispatched based on native_functions.yaml?

I'm sure there are many ways we could make the promotion change, but, I think it's generally preferable to "fail fast," i.e., better to have the error checking code continue to happen before we perform any promotion operations. I also think it's preferable to have the can_cast check happen within the exiting error checking code, since it's related. I'd rather not do too much refactoring right now in the way dispatching happens, and `add` does currently have four types of calls.

The changes in the PR as-written are generally just to get and check the common dtype as part of "1", call `.to(commonDtype)` when doing the values extraction in step 2, reference the commonDtype instead of `r.scalar_type()` while performing the op, and modify step 4 to convert the common-dtype result back to the dtype of the result tensor (if necessary).

To the extent I deviated from that, it's for op-specific reasons. e.g. for mixed-type addition, I felt like it was probably important to call coalesce before down-casting back to the return type even though for same-type ops we don't.

If promotion could all go in one isolated function (for all ops) that would be helpful, but I don't think it's so meaty that it deserves another unique layer of indirection in every call site, and we don't have something like TensorIterator (or even universal support for all ops) for sparse tensors.

We _could_ wrap step 2 and 3 above into a separate function (for each op), calling `to` on the tensors passed to that function instead of on the values extracted... but that doubles the number of functions and I'm not sure it improves readability. It wouldn't put the promotion logic all in one place, still scattering it across all (existing) functions.

We could make sparse math support more universal, which I think could simplify this code too since then each op would have the same error cases etc, but that seems like it'd be a different project.

Reply:
>> a function like add has ~4 different places that do type promotion separately.

> I think this just means the four calls (add_sparse, add_sparse_, add_out_sparse_cpu, and add_out_sparse_cuda) dispatched based on native_functions.yaml?

I mean `add_out_sparse_non_contiguous`, `add_out_sparse_contiguous`, `add_out_dense_sparse_cpu`, `add_sparse`,

Reply:
ah I see, for `add` where I split the implementation code into add_out_sparse_contiguous/non_contiguous, some promotion code leaked into the implementing functions. The conversion could probably be pulled back into the calling function somehow, but they are slightly different, and pulling it out would make add_out_sparse_cpu more complicated, and it still needs to return references so the implementing functions would still have to do the kind of `set_indices...` calls that they do.